payroll/models/hr_payslip.py

```diff
@@ -587,117 +466,127 @@ def get_contract_dict(self, contract, contracts):
 
         return {}
 
-    @api.model
-    def _get_payslip_lines(self, contract_ids, payslip_id):
-        def _sum_salary_rule_category(localdict, category, amount):
-            if category.parent_id:
-                localdict = _sum_salary_rule_category(
-                    localdict, category.parent_id, amount
-                )
-
-            if category.code in localdict["categories"].dict:
-                localdict["categories"].dict[category.code] += amount
-            else:
-                localdict["categories"].dict[category.code] = amount
-
-            return localdict
-
-        payslip = self.env["hr.payslip"].browse(payslip_id)
-        contracts = self.env["hr.contract"].browse(contract_ids)
-
-        # we keep a dict with the result because a value can be overwritten by
-        # another rule with the same code
-        result_dict = {}
-        rules_dict = {}
-        contract_dict = {}
-        blacklist = []
-        categories = BrowsableObject(payslip.employee_id.id, {}, self.env)
-        rules = BrowsableObject(payslip.employee_id.id, rules_dict, self.env)
-        current_contract = BrowsableObject(
-            payslip.employee_id.id, contract_dict, self.env
-        )
-
-        baselocaldict = payslip.get_baselocaldict(contracts)
-        baselocaldict["categories"] = categories
-        baselocaldict["rules"] = rules
-        baselocaldict["current_contract"] = current_contract
+    def _get_baselocaldict(self, contracts):
+        self.ensure_one()
+        worked_days_dict = {
+            line.code: line for line in self.worked_days_line_ids if line.code
+        }
+        input_lines_dict = {
+            line.code: line for line in self.input_line_ids if line.code
+        }
+        localdict = {
+            **{
+                "payslip": Payslips(self.employee_id.id, self, self.env),
+                "worked_days": WorkedDays(
+                    self.employee_id.id, worked_days_dict, self.env
+                ),
+                "inputs": InputLine(self.employee_id.id, input_lines_dict, self.env),
+                "payroll": BrowsableObject(
+                    self.employee_id.id, self.get_payroll_dict(contracts), self.env
+                ),
+                "current_contract": BrowsableObject(self.employee_id.id, {}, self.env),
+                "categories": BrowsableObject(self.employee_id.id, {}, self.env),
+                "rules": BrowsableObject(self.employee_id.id, {}, self.env),
+                "result_rules": BrowsableObject(self.employee_id.id, {}, self.env),
+            },
+        }
+        return localdict
```

Comment:
The following keys either already have some way to be overridden or they are odoo models that can be directly modified by the developer:
"payslip", "worked_days", "inputs", "payroll", "current_contract"
The "categories"  and "rules" keys are there to be modified by the salary rule calculation code so if someone wants to mess with them the proper thing for them to do is inherit the salary calculation code.

You took what was a relatively straight-forward method of injecting arbitrary values into to be used **by** the salary rules and it seems you tried to turn it into a general mechanism for overriding the calculations themselves but in the process you made it overly complicated. 

Also, do you have a reason for making this change? And I don't mean a hypothetical "it may be useful in the future" reason, but an "I need it because I'm doing so and so and this is the only way I can thing of to achieve it' reason. If not then I would suggest to leave this piece of code alone.

Reply:
Actually the reason for it is that you can override them without touching the calculation loop which is what i intended to do. 

I use this for example to use localdict context in other places like a module that simulates payslip calculation and actually in a module for profit calculation that needs to inject values in the localdict to be summed with current calculation values and then writed to a table. 

With current behavior if you want to create or edit dicts inside the salary rules, you can do it (it's supported) but you then don't have a way to take that values when the calculation loop ends. 

That's why localdict is also a returned value, because I have already two implementations that require doing calculations on the fly when the rules are being calculated. If we only return the lines, we can't get any other data out the loop. For easy payroll structured this is okay but at least in the implementations that i had to work on, we need to update tables on the fly to get updated values that are dependencies of other rules. 

I know it seems complicated but I've seen code like this in other modules of odoo. 

I invite you to try it and you will see it opens a new whole of new possibilities to allow to inject values, and also it makes no harm because they still are all private functions. 

Then I can show you in my private repo if you want to see how this is useful, really, I think you would like it because it's very similar to your payroll and current_contract dict approach but allows a new level of customization to happen in salary rules.